File: SenLib/Sen2/FileFixes/scripts_book_dat_us_book04_dat.cs
using HyoutaUtils;
using System.Collections.Generic;
using System.IO;

namespace SenLib.Sen2.FileFixes {
	public class scripts_book_dat_us_book04_dat : FileMod {
		public string GetDescription() {
			return "Fix translation error in headline of Imperial Chronicle Issue 3.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x5b2fee612159bcb9ul, 0x3b2c6831f94f7b1ful, 0x4dd6231cu));
			if (s == null) {
				return null;
			}
			MemoryStream bin = s.CopyToMemoryAndDispose();

			// translation error in last headline of imperial chronicle 3
			bin.Position = 0x4124;
			bin.WriteUInt16(0x4561, EndianUtils.Endianness.BigEndian);

			return new FileModResult[] { new FileModResult("data/scripts/book/dat_us/book04.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x5b2fee612159bcb9ul, 0x3b2c6831f94f7b1ful, 0x4dd6231cu));
			if (s == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_us/book04.dat", s) };
		}
	}
}

File: native/sen2/file_fixes/scripts_book_dat_us_book04_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "util/hash/sha1.h"
#include "util/memwrite.h"

namespace SenLib::Sen2::FileFixes::scripts_book_dat_us_book04_dat {
std::string_view GetDescription() {
    return "Fix translation error in headline of Imperial Chronicle Issue 3.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/book/dat_us/book04.dat",
            51414,
            HyoutaUtils::Hash::SHA1FromHexString("5b2fee612159bcb93b2c6831f94f7b1f4dd6231c"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        // translation error in last headline of imperial chronicle 3
        HyoutaUtils::MemWrite::WriteArray(&bin[0x4124], {{0x45, 0x61}});

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen2::FileFixes::scripts_book_dat_us_book04_dat
